src/datasets/formatting/formatting.py

```diff
@@ -221,6 +230,9 @@ def format_column(self, pa_table: pa.Table) -> pd.Series:
     def format_batch(self, pa_table: pa.Table) -> pd.DataFrame:
         return self.pandas_arrow_extractor().extract_batch(pa_table)
 
+    def format_batch_with_category(self, pa_table: pa.Table, features: Features) -> pd.DataFrame:
```

Comment:
@lhoestq - this was unfortunately pretty messy, so I'd love to get your thoughts!

The biggest problem I ran into is that on the existing method, when pyarrow is being converted back to pandas, we don't have access to the ClassLabel, which is what holds the actual names.

Even if we added features as an optional argument to the formatter base methods, the formatters are invoked in places that don't actually have a Features() present in the context (e.g. in `format_table`)

So I added another method and used this in `.to_pandas()` but this doesn't get used for the other pandas conversion points and I don't like that there's inconsistent behavior between these :-(

Would love to get your thoughts, but at the moment I'm kind of leaning towards leaving `.from_pandas()` working for Category -> ClassLabel, and just letting `.to_pandas()` be ClassLabel -> int (which would also be consistent with the numpy, python, etc. formats).

Reply:
What about adding `features` in the `__init__` of the formatter ? This way you can have `self.features` for the formatter and get the label names if needed.

Let me know what you think ? IMO it makes sense that the formatter knows the features
I can make a separate PR to first add the `features` accessible as an attribute of the formatter if it can help

Reply:
I think that would make this much more tractable - I took a look at the impl for that, and it would help quite a bit if you would be able to do that (but it's not a super high priority so don't worry about it if you can't get to it).

Reply:
I'll do that in a few days, I've been pretty busy sorry ;)

Reply:
Please no apologies!  It's totally no worries - we've monkeypatched this in our codebase, so it's not blocking us at all - so whenever it's convenient for you and fits with your priorities (and if you end up deciding it's not worth it, that's totally fine too!)

I should be thanking you for working on this software that we're using! :-)